### PR TITLE
add whoami endpoint

### PR DESCRIPTION
**Pull request by jrwren**

### CODE REVIEW

**Comment by jrwren** [comment]
WIP: ~~I try to QA this with identity on localhost:8081 and charmstore on localhost:8080 and after I create an agent in identity, I get `cannot get groups for evarlast: user does not have correct permissions` and I'm not sure why blues-identity is not giving the agent the ability to read groups. Anyone have ideas?~~ I didn't add the agent to any groups. Once I added it to admin@idm group, it worked.

QA:
1. run local charmstore(this branch) and local identity, make sure charmstore is correctly configured for identity agent auth
2. hit the /whoami endpoint e.g. `http localhost:8080/v4/whoami` If you are already logged in you will see a list of groups. If you are not logged in you with be prompted for auth.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1097/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1098/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1099/
Test FAILed.

File: internal/v4/api_test.go [hunk 1]
```diff
@@ -2206,6 +2206,37 @@ func (s *APISuite) TestMacaroon(c *gc.C) {
...
+func (s *APISuite) TestWhoAmIReturnsNameAndGroups(c *gc.C) {
+	s.discharge = dischargeForUser("who")
+	s.idM.groups = map[string][]string{
+		"who": []string{"foo", "bar"},
+	}
+	httptesting.AssertJSONCall(c, httptesting.JSONCallParams{
+		Handler: s.srv,
+		URL:     storeURL("whoami"),
+		Do:      bakeryDo(nil),
+		//		Cookies:      []*http.Cookie{macaroonCookie},
```
Comment (mhilton) [inline internal/v4/api_test.go]: Please remove commented out code.

**Comment by mhilton** [comment]
:+1: LGTM, FWIW the test failure looks spurious.

File: docs/API.md [hunk 2]
```diff
@@ -1806,6 +1806,19 @@ A delegatable macaroon will only be returned to an authorized user (not
 including admin). It will carry the same privileges as the macaroon used
 to authorize the request.
 
+#### GET /whoami
+
+This endpoint returns a user name and list of groups in JSON format.
```
Comment (rogpeppe) [inline docs/API.md]: This endpoint returns the user name of the client and the list of groups the user
is a member of. This endpoint requires authorization.

?

(no need to mention JSON format - it's implicit in all the endpoints in this doc)

Comment (rogpeppe) [inline docs/API.md]: also:

```
type WhoAmIResponse struct {
    User string
    Groups []string
}
```

File: docs/API.md [hunk 3]
```diff
@@ -1806,6 +1806,19 @@ A delegatable macaroon will only be returned to an authorized user (not
...
+
+Example: `GET whoami`
+
+```json
+{
+    "User": ["alice"],
```
Comment (rogpeppe) [inline docs/API.md]: Just "alice", I think.

File: internal/v4/api.go [hunk 4]
```diff
@@ -1011,6 +1012,26 @@ func (h *ReqHandler) serveDelegatableMacaroon(_ http.Header, req *http.Request)
...
 }
 
+// GET /whoami
+// See https://github.com/juju/charmstore/blob/v4/docs/API.md#whoami
+func (h *ReqHandler) serveWhoAmI(_ http.Header, req *http.Request) (interface{}, error) {
+	auth, err := h.authorize(req, []string{params.Everyone}, true, nil)
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Any)
+	}
+	if auth.Username == "" {
```
Comment (rogpeppe) [inline internal/v4/api.go]: if auth.Admin {

?

**Comment by rogpeppe** [comment]
LGTM

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1103/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1104/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1105/
Test PASSed.

**Comment by jrwren** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**